### PR TITLE
Track storage gas consumed

### LINKED ISSUE #8: Count storage gas cost
Keep track of how much gas is used for storage so that we can get cpu_cost and storage_cost easily

### LINKED ISSUE #10: Allow set/delete pair to reclaim all storage costs
- The reclaim must happen as you go instead of happening in the end
- Ideally, if the user deletes first then stores, they should be able to do that without have a high gas_limit as reserve
- But the above mechanism should not allow the user to perform more computation using the gas quota released from delete
- (Maybe) The delete/set should be allowed in function calls. (e.g. if C calls A first to release some storage then C calls B to store something, C should be able to take advantage of the above mechanism)

### PR DESCRIPTION
**Pull request by elmato**

### CODE REVIEW

File: lib/evmone/advanced_analysis.cpp [hunk 1]
```diff
@@ -92,12 +84,7 @@ AdvancedCodeAnalysis analyze(evmc_revision rev, bytes_view code, const AdvancedE
         block.stack_req = std::max(block.stack_req, opcode_info.stack_req - block.stack_change);
         block.stack_change += opcode_info.stack_change;
         block.stack_max_growth = std::max(block.stack_max_growth, block.stack_change);
-
-        if(state.eos_evm_version > 0 && (opcode == OP_CREATE || opcode == OP_CREATE2)) {
-            block.gas_cost += static_cast<int64_t>(state.gas_params.G_txcreate);
-        } else {
-            block.gas_cost += opcode_info.gas_cost;
-        }
+        block.gas_cost += opcode_info.gas_cost;
```
Comment (yarkinwho) [inline lib/evmone/advanced_analysis.cpp:87]: Does this part of code mean we fixed the gas_cost so we do not need to special case OP_CREATE and OP_CREATE2 ?

File: lib/evmone/instructions_calls.cpp [hunk 2]
```diff
@@ -139,6 +154,17 @@ Result create_impl(StackTop stack, int64_t gas_left, ExecutionState& state) noex
...
 
+    // OP_CREATE/OP_CREATE2 gas cost (32000) is constant among all evmc revisions up to Cancun
+    int64_t gas_cost = 32000;
+    if(state.eos_evm_version >= 3) {
+        gas_cost = state.gas_state.apply_storage_gas_delta(static_cast<int64_t>(state.gas_params.G_txcreate));
+    } else if (state.eos_evm_version >= 1) {
+        gas_cost = static_cast<int64_t>(state.gas_params.G_txcreate);
+    }
+
+    if (INTX_UNLIKELY((gas_left -= gas_cost) < 0))
```
Comment (yarkinwho) [inline lib/evmone/instructions_calls.cpp:165]: Is this unlikely proper here? as estimate gas should constantly hit true here.

File: lib/evmone/instructions_traits.hpp [hunk 3]
```diff
@@ -100,7 +100,7 @@ constexpr inline GasCostTable gas_costs = []() noexcept {
         table[EVMC_FRONTIER][op] = 3;
     for (auto op = size_t{OP_LOG0}; op <= OP_LOG4; ++op)
         table[EVMC_FRONTIER][op] = static_cast<int16_t>((op - OP_LOG0 + 1) * 375);
-    table[EVMC_FRONTIER][OP_CREATE] = 32000;
+    table[EVMC_FRONTIER][OP_CREATE] = 0;
```
Comment (yarkinwho) [inline lib/evmone/instructions_traits.hpp:103]: What's the logic behind this?

File: lib/evmone/advanced_analysis.cpp (hunk 1 above)
Comment (elmato) [reply lib/evmone/advanced_analysis.cpp:87]: We reverted this part of the code analysis to how it was originally done in evmone.
What we are doing instead is setting the cost of CREATE and CREATE2 to 0 in the original static tables of evmone and we are making the cost dynamic in the create (create_impl) instruction implementation.

Comment (elmato) [reply lib/evmone/advanced_analysis.cpp:87]: https://github.com/eosnetworkfoundation/evmone/pull/12/files/12e07e659b8f93dda13516ba0cf790e2537eedec#diff-8dcb083e9ac17ef66c94c342ae2fa8efc85ecb89e9794381780c333805443f57R157-R165

File: lib/evmone/instructions_traits.hpp (hunk 3 above)
Comment (elmato) [reply lib/evmone/instructions_traits.hpp:103]: See the comment above

**Review by yarkinwho — APPROVED** [review]

**Pull request merged**